core/test/unit/models/plugins/filter_spec.js

```diff
@@ -24,639 +24,778 @@ describe('Filter', function () {
+                        {tag:'photo'},
+                        {page:false}
+                    ]};
+
+                    mergeFilters(input).should.eql(output);
+                });
+
+                it('should reduce default filters if default and custom are same', () => {
+                    const input = {
+                        defaults: {page:false},
+                        custom: {page:true},
+                    };
+                    const output = {page:true};
+
+                    mergeFilters(input).should.eql(output);
+                });
+
+                it('should match nested $and with a key inside primary filter', function () {
+                    const input = {
+                        defaults: {
+                            $and:[
+                                {page:false},
+                                {status: 'published'}
+                            ]
+                        },
+                        custom: {
+                            page: {
+                                $in:[false,true]
+                            }
+                        }
+                    };
+                    const output = {$and: [
+                        {page: {
+                            $in:[false,true]
+                        }},
+                        {$and:[
+                            {status: 'published'},
+                        ]}
+                    ]};
+
+                    mergeFilters(input).should.eql(output);
+                });
+
+                it('should reduce default filters if default and custom overlap', () => {
+                    const input = {
+                        defaults: {$or:[
+                            {page:false},
+                            {author:'cameron'}
+                        ]},
+                        custom: {$and: [
+                            {tag: 'photo'},
+                            {page: true}
+                        ]}
+                    };
+                    const output = {$and:[
+                        {$and: [
+                            {tag:'photo'},
+                            {page:true},
+                        ]},
+                        {$or: [
+                            {author:"cameron"}
+                        ]}
+                    ]};
+
+                    mergeFilters(input).should.eql(output);
+                });
+
+                it('should return a merger of enforced and defaults plus custom filters if provided', () => {
+                    const input = {
+                        enforced: {status:'published'},
+                        defaults: {page:false},
+                        custom: {tag:'photo'},
+                    };
+                    const output = {$and: [
+                        {$and: [
+                            {status:'published'},
+                            {tag:'photo'}
+                        ]},
+                        {page:false}
+                    ]};
```

Comment:
Good observation. At the moment the grouping logic is very simple and doesn't recognize if the group would produce an equivalent result like in this example. I see it more like an optimization that could be done later. But maybe there's some use case I'm missing? :thinking:

Reply:
I have not really looked in detail at the logic. IMO the logic does too many $and connections. And i think it might be hard to detect this on runtime, because you don't know the values of the filters.

I thought the logic works like this:

You receive:

- enforced (0)
- custom (1)
- default (2)


You reject any key of `default` which is in all the other filters.
You repeat that for custom.

Now you combine all three together ($and: [enforced, custom, default]).

This would also simplify the logic a lot and we would only connect $and once.

Reply:
There is a big discussion of this logic here: https://github.com/NexesJS/nql/issues/2#issuecomment-442038597

IMO you have to work through the cascade to get the right logic. 

Having outer AND statements ensures safety. We can look to optimise this later, but IMO this should not block shipping as the current impl is safer, even if it is not optimal.